power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -21,114 +25,242 @@ namespace detail = power_grid_model::optimizer::detail;
+    edges.reserve(std::distance(iter.begin(), iter.end()) * 2);
+    edge_props.reserve(std::distance(iter.begin(), iter.end()) * 2);
+    for (auto const& branch : iter) {
+        if (!branch.from_status() || !branch.to_status()) {
+            continue;
+        }
+        add_to_edge(edges, edge_props, branch.from_node(), branch.to_node(), {unregulated_idx, 0});
+        add_to_edge(edges, edge_props, branch.to_node(), branch.from_node(), {unregulated_idx, 0});
+    }
+}
+
+template <typename... ComponentTypes, main_core::main_model_state_c State>
+inline auto add_edges(State const& state, RegulatedObjects const& regulated_objects, TrafoGraphEdges& edges,
+                      TrafoGraphEdgeProperties& edge_props) {
+    (add_edge<ComponentTypes>(state, regulated_objects, edges, edge_props), ...);
+}
+
+inline void add_to_edge(TrafoGraphEdges& edges, TrafoGraphEdgeProperties& edge_props, Idx const& start, Idx const& end,
+                        TrafoGraphEdge const& edge_prop) {
+    edges.emplace_back(start, end);
+    edge_props.emplace_back(edge_prop);
+}
+
 template <main_core::main_model_state_c State>
-inline auto build_transformer_graph(State const& /*state*/) -> TransformerGraph {
-    // TODO(nbharambe): implement
-    return {};
+inline auto retrieve_regulator_info(State const& state) -> RegulatedObjects {
+    RegulatedObjects regulated_objects;
+    for (auto const& regulator : state.components.template citer<TransformerTapRegulator>()) {
+        if (!regulator.status()) {
+            continue;
+        }
+        if (regulator.regulated_object_type() == ComponentType::branch) {
+            regulated_objects.transformers.emplace(regulator.regulated_object());
+        } else {
+            regulated_objects.transformers3w.emplace(regulator.regulated_object());
+        }
+    }
+    return regulated_objects;
+}
+
+template <main_core::main_model_state_c State>
+inline auto build_transformer_graph(State const& state) -> TransformerGraph {
+    TrafoGraphEdges edges;
+    TrafoGraphEdgeProperties edge_props;
+
+    const RegulatedObjects regulated_objects = retrieve_regulator_info(state);
+
+    add_edges<Transformer, ThreeWindingTransformer, Line, Link>(state, regulated_objects, edges, edge_props);
+
+    // build graph
+    TransformerGraph trafo_graph{boost::edges_are_unsorted_multi_pass, edges.cbegin(), edges.cend(),
+                                 edge_props.cbegin(),
+                                 static_cast<TrafoGraphIdx>(state.components.template size<Node>())};
+
+    BGL_FORALL_VERTICES(v, trafo_graph, TransformerGraph) { trafo_graph[v].is_source = false; }
+
+    if (state.components.template size<Source>() == 0) {
+        return trafo_graph;
+    }
+
+    // Mark sources
+    for (auto const& source : state.components.template citer<Source>()) {
+        // ignore disabled sources
+        trafo_graph[source.node()].is_source = source.status();
+    }
```

Comment:
overoptimization: if the amount of sources is 0 the `for` loop doesn't do anything anyway. (in fact, it introduces an unnecessary branch so it's actually even slower in almost every case)

```suggestion
    // Mark sources
    for (auto const& source : state.components.template citer<Source>()) {
        // ignore disabled sources
        trafo_graph[source.node()].is_source = source.status();
    }
```

Reply:
The `citer` not work if the component does not exist. Hence the size check is added first.

Reply:
in my branch, i added this constexpr check: https://github.com/PowerGridModel/power-grid-model/pull/557/files#diff-1d8601478861bbe22e22c742aa5db8c51c0fe1fa9c5a225193dcb72cce714ef8R52-R53

i think you can combine that with adding to the container something like

```cpp
    // type checks
    template <typename... T>
    static constexpr has_gettable_type = supported_type_c<T, GettableTypes...>;
    template <typename... T>
    static constexpr has_storageable_type = supported_type_c<T, StorageableTypes...>;
```

Reply:
i can cherry-pick these changes if you like

Reply:
Then actually we dont need those size checks too ;(

Reply:
do you want me to or not? hahaha

Reply:
Yes please!